hasktorch/test/PipelineSpec.hs

```diff
@@ -0,0 +1,61 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE FlexibleContexts #-}
+module PipelineSpec where
+
+import Control.Concurrent
+import Control.Concurrent.Async
+import Control.Monad
+import Pipes
+import Pipes.Concurrent
+import System.Exit
+import System.IO
+import System.Timeout
+import Torch.Data.Pipeline
+  
+import Test.Hspec
+
+
+defaultTimeout :: Int
+defaultTimeout = 100000  
+
+newtype MockData = MockData Int
+instance Dataset MockData Int where
+  getBatch (MockData iters) _ = pure $ iters
+  numIters (MockData iters)  = iters
+
+instance ConcurrentDataset MockData Int where
+  getBatchConcurrently _ = getBatch 
+
+testFoldTimeout :: MockData -> IO ()
+testFoldTimeout dataset = do
+  (toBatches, fromBatches, sealBatch) <- spawn' (bounded 1)
+  thread <- async $ runEffect $ readBatches dataset toBatches
+  fold <- pure $ foldFromProducer (takeBatch fromBatches)
+  timedOut <- async $ fold $ FoldM takeBatchThenTimeout (pure 0) pure
+  wait thread
+  cancel timedOut
+
+
+testConcurrentFoldTimeout :: MockData  -> Int -> IO ()
```

Comment:
I'm trying to make sense of this. Can you maybe add some documentation that explains the expected behaviour?
I've come up with this:

working: `-`
idling:  `.`

#### with 1 worker and batching as slow as folding:

```
mailbox:  00000010000010000
worker 0: |-----|-----|----
fold:     |.....|-----|----
```

#### with 2 workers and batching twice as slow as folding:

```
mailbox:  00000021100010010
worker 0: |-----|-----|----
worker 1: |-----|..|-----|-
fold:     |.....|--|--|--|-
```